### PR TITLE
test: e2e test chash upstream hash on (consumer, header, cookie)  

### PR DESCRIPTION
**Pull request by idbeta**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] E2E test
- [ ] New feature provided
- [ ] Improve performance

- Related issues

related issue: https://github.com/apache/apisix-dashboard/issues/908

### CODE REVIEW

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+		resp, err = http.DefaultClient.Do(req)
+		assert.Nil(t, err)
+		respBody, err = ioutil.ReadAll(resp.Body)
+		body := string(respBody)
+		if _, ok := res[body]; !ok {
+			res[body] = 1
+		} else {
+			res[body] += 1
+		}
+	}
```
Comment (nic-chen) [inline api/test/e2e/upstream_chash_hash_on_test.go:95]: we could put them into a function and reuse it

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+	time.Sleep(time.Duration(100) * time.Millisecond)
+	basepath := "http://127.0.0.1:9080"
+	var req *http.Request
+	var err error
+	var url string
+	var resp *http.Response
+	var respBody []byte
+	res := map[string]int{}
+	for i := 0; i <= 3; i++ {
+		url = basepath + "/server_port?var=2&var2=" + strconv.Itoa(i)
```
Comment (nic-chen) [inline api/test/e2e/upstream_chash_hash_on_test.go:154]: here is testing cookie ? why has args ?

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+			res[body] += 1
+		}
+	}
+	assert.Equal(t, true, res["1980"] == 4 || res["1981"] == 4)
+	resp.Body.Close()
+
+	//hit routes with miss cookie
+	res = map[string]int{}
+	for i := 0; i <= 3; i++ {
+		url = basepath + "/server_port?var=2&var2=" + strconv.Itoa(i)
```
Comment (nic-chen) [inline api/test/e2e/upstream_chash_hash_on_test.go:173]: ditto

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+		resp, err = http.DefaultClient.Do(req)
+		assert.Nil(t, err)
+		respBody, err = ioutil.ReadAll(resp.Body)
+		body := string(respBody)
+		if _, ok := res[body]; !ok {
+			res[body] = 1
+		} else {
+			res[body] += 1
+		}
+	}
```
Comment (nic-chen) [inline api/test/e2e/upstream_chash_hash_on_test.go:185]: could use a same function with above cases.

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
```
Comment (nic-chen) [inline api/test/e2e/upstream_chash_hash_on_test.go:256]: reuse.

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
```
Comment (nic-chen) [inline api/test/e2e/upstream_chash_hash_on_test.go:334]: ditto

File: api/test/e2e/upstream_chash_hash_on_test.go (hunk 2 above)
Comment (idbeta) [reply api/test/e2e/upstream_chash_hash_on_test.go]: fixed.

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+			ExpectStatus: http.StatusOK,
+			Sleep:        sleepTime,
+		},
+	}
+
+	for _, tc := range tests {
+		testCaseCheck(tc)
+	}
+
+	//hit routes
```
Comment (membphis) [inline api/test/e2e/upstream_chash_hash_on_test.go:73]: style

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+						"weight": 1
+					}],
+					"type": "chash",
+					"key": "custom-cookie",
+					"hash_on": "cookie"
+				}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+		},
+		{
```
Comment (membphis) [inline api/test/e2e/upstream_chash_hash_on_test.go:125]: I think we can delete this step, same as: https://github.com/apache/apisix-dashboard/pull/936/files#diff-d07aed971e7a3caa6a1ae0527badd308d4be4053dafaa873aec158761b6c426cR55

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+					}],
+					"type": "chash",
+					"key": "X-Sessionid",
+					"hash_on": "header"
+				}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+		},
+		{
+			caseDesc: "create route using the upstream just created",
```
Comment (membphis) [inline api/test/e2e/upstream_chash_hash_on_test.go:216]: ditto

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+		assert.Nil(t, err)
+		respBody, err = ioutil.ReadAll(resp.Body)
+		body := string(respBody)
+		if _, ok := res[body]; !ok {
+			res[body] = 1
+		} else {
+			res[body] += 1
+		}
+	}
+	assert.Equal(t, true, res["1980"] == 4 || res["1981"] == 4)
```
Comment (membphis) [inline api/test/e2e/upstream_chash_hash_on_test.go:186]: need some comment, it is not easy to understand

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+		assert.Nil(t, err)
+		respBody, err = ioutil.ReadAll(resp.Body)
+		body := string(respBody)
+		if _, ok := res[body]; !ok {
+			res[body] = 1
+		} else {
+			res[body] += 1
+		}
+	}
+	assert.Equal(t, true, res["1980"] == 8 && res["1981"] == 8)
```
Comment (membphis) [inline api/test/e2e/upstream_chash_hash_on_test.go:257]: ditto

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+		assert.Nil(t, err)
+		respBody, err = ioutil.ReadAll(resp.Body)
+		body := string(respBody)
+		if _, ok := res[body]; !ok {
+			res[body] = 1
+		} else {
+			res[body] += 1
+		}
+	}
+	assert.Equal(t, true, res["1980"] == 4 || res["1981"] == 4)
```
Comment (membphis) [inline api/test/e2e/upstream_chash_hash_on_test.go:337]: ditto

File: api/test/e2e/upstream_chash_hash_on_test.go (hunk 4 above)
Comment (idbeta) [reply api/test/e2e/upstream_chash_hash_on_test.go:185]: I will improve this in the future.

File: api/test/e2e/upstream_chash_hash_on_test.go (hunk 8 above)
Comment (idbeta) [reply api/test/e2e/upstream_chash_hash_on_test.go]: fixed.

File: api/test/e2e/upstream_chash_hash_on_test.go [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
```
Comment (membphis) [inline api/test/e2e/upstream_chash_hash_on_test.go:97]: need a comment too

**Review by membphis — APPROVED** [review]
LGTM, except minor things

File: api/test/e2e/upstream_chash_hash_on_test.go (hunk 1 above)
Comment (idbeta) [reply api/test/e2e/upstream_chash_hash_on_test.go:96]: I add it to the todo list.

**Review by nic-chen — APPROVED** [review]
LGTM Except for some code format issues

**Pull request merged**